File: app/src/main/java/com/example/mynotes/NoteTakerActivity.java
package com.example.mynotes;

import androidx.appcompat.app.AppCompatActivity;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import com.example.mynotes.model.Note;

import java.text.SimpleDateFormat;
import java.util.Date;

public class NoteTakerActivity extends AppCompatActivity {
    EditText editText_title,editText_description;
    ImageView imageView_save;
    Note note;
    Boolean isOldNote = false;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_note_taker);
        editText_title = findViewById(R.id.editText_title);
        editText_description = findViewById(R.id.editText_description);
        imageView_save = findViewById(R.id.imageView_save);
        note = new Note();
        try {
            note = (Note)getIntent().getSerializableExtra("old_note");
            editText_title.setText(note.getTitle());
            editText_description.setText(note.getDescription());
            isOldNote = true;
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
        imageView_save.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String title = editText_title.getText().toString();
                String description = editText_description.getText().toString();
                if(description.isEmpty())
                {
                    Toast.makeText(NoteTakerActivity.this, "Please add some note !", Toast.LENGTH_SHORT).show();
                    return;
                }

                SimpleDateFormat formatter = new SimpleDateFormat("EEE, d MMM yyyy HH:mm a");
                Date date = new Date();

                if(!isOldNote)
                {
                    note = new Note();
                }

                note.setTitle(title);
                note.setDescription(description);
                note.setDate(formatter.format(date));
                Intent intent = new Intent();
                intent.putExtra("note",note);
                setResult(Activity.RESULT_OK,intent);
                finish();

            }
        });
    }
}

File: app/src/main/java/com/example/mynotes/NoteTakerActivity.kt
package com.example.mynotes

import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import com.example.mynotes.databinding.ActivityNoteTakerBinding
import com.example.mynotes.model.Note
import com.example.mynotes.viewmodel.NoteViewModel
import com.example.mynotes.viewmodel.NoteViewModelFactory
import com.google.android.material.dialog.MaterialAlertDialogBuilder

class NoteTakerActivity : AppCompatActivity() {

    private val viewModel: NoteViewModel by viewModels {
        NoteViewModelFactory(
            (application as NoteApplication).database.noteDao(),
        )
    }

    lateinit var note: Note
    private var _binding: ActivityNoteTakerBinding? = null
    private val binding get() = _binding!!
    private var oldNoteId = 0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = ActivityNoteTakerBinding.inflate(layoutInflater)
        setContentView(binding.root)
        oldNoteId = intent.getIntExtra("old_note_id",0)
        if(oldNoteId>0)
        {
            binding.imageViewDelete.visibility = View.VISIBLE
            viewModel.retrieveNote(oldNoteId).observe(this) { selectedNote ->
                note = selectedNote
                bind(note)
            }
        } else {
            binding.imageViewSave.setOnClickListener {
                addNewNote()
            }
        }

    }

    private fun bind(note: Note) {
        binding.apply {
            editTextTitle.setText(note.title)
            editTextDescription.setText(note.description)
            imageViewSave.setOnClickListener {
                updateNote()
            }
            imageViewDelete.setOnClickListener {
                showConfirmationDialog()
            }
        }
    }

    private fun addNewNote() {
        if (isEntryValid()) {
            viewModel.addNewNote(
                binding.editTextTitle.text.toString(),
                binding.editTextDescription.text.toString(),
            )
            finish()
        }
    }

    private fun updateNote() {
        if (isEntryValid()) {
            viewModel.updateNote(
                oldNoteId,
                binding.editTextTitle.text.toString(),
                binding.editTextDescription.text.toString(),
            )
            finish()
        }
    }

    private fun isEntryValid(): Boolean {
        return viewModel.isNoteValid(
            binding.editTextTitle.text.toString(),
            binding.editTextDescription.text.toString()
        )
    }

    private fun showConfirmationDialog() {
        MaterialAlertDialogBuilder(this)
            .setTitle(getString(android.R.string.dialog_alert_title))
            .setMessage(getString(R.string.delete_question))
            .setCancelable(false)
            .setNegativeButton(getString(R.string.no)) { _, _ -> }
            .setPositiveButton(getString(R.string.yes)) { _, _ ->
                deleteNote()
            }
            .show()
    }

    private fun deleteNote() {
        viewModel.deleteNote(note)
    }


}